### DISCUSSION TITLE
Namespace & Names Case Sensitive

### DISCUSSION THREAD

**Discussion by calvinrp** [bytecodealliance/registry#231] [General]
From my read of the Component Model spec and the validation code, namespaces just need to validate as the `KebabStr` (see [validation code](https://github.com/bytecodealliance/wasm-tools/blob/main/crates/wasmparser/src/validator/names.rs#L52-L72)).

So `my-NAME`, `MY-NAME`, `MY-name` and `my-name` are all valid `KebabStr`. I assume we would want to match case insensitive to prevent confusion.

Should namespaces be required to be lowercased to pass validation? Or just match case-insensitive? If not lowercased, do we want to keep the original casing for presentation purposes?

On a similar note, how do we want to handle casing in package names with respect to the registry and `LogId`s? It is possible to hash as lowercase. Currently, `PackageId` (soon to be `PackageName`) is case sensitive and produces different `LogId`s depending on the case.

**Comment by lukewagner** [comment]
My impression is that we should be case sensitive, but disallow two package names that differ only by case.  Thus, we're never throwing away case (which is meaningful, since it renders into source code identifiers differently), just preventing ambiguities or possibly, depending on the casing scheme, collisions.  (Maybe the component model should do likewise in its uniqueness...)

**Reply by calvinrp** [reply]
Thanks, I was thinking the same. Case preserved for presentation. But enforce uniqueness and `LogId` hashing in a case insensitive way.

**Reply by lann** [reply]
Downcasing when hashing for the LogID should effectively enforce case-insensitive uniqueness.

**Reply by calvinrp** [reply]
> Downcasing when hashing for the LogID should effectively enforce case-insensitive uniqueness.

Yep, I was thinking the same. Modify the code to lowercase when hashing in the `PackageId` / `PackageName` for `LogId`.

**Reply by esoterra** [reply]
I would rather we reject namespace names which are mixed case or uppercase so that we don't have log data with mixed upper/lower and propagate an error to users that the namespace they chose is not legal.

**Reply by lann** [reply]
Lower-only namespaces seems like a pretty reasonable starting point but I'm a bit wary of diverging from WIT...

**Reply by calvinrp** [reply]
> I would rather we reject namespace names which are mixed case or uppercase so that we don't have log data with mixed upper/lower and propagate an error to users that the namespace they chose is not legal.

We could allow case sensitive names and namespaces for presentation and code gen. But generate the `LogId` in a case insensitive way (all lowercase).

**Reply by lann** [reply]
I guess if namespaces definition is generally considered to be the responsibility of a registry then lowercase-only is really just policy.

**Reply by calvinrp** [reply]
> Lower-only namespaces seems like a pretty reasonable starting point but I'm a bit wary of diverging from WIT...

Yeah, I'm concerned about that. Think we need to support what is valid in the component.

**Reply by esoterra** [reply]
@lukewagner are we expecting namespace names to be full kebab names? I hadn't seen any examples of multi-part namespace before and assumed they were just all-lower one part stubs?

If they're full kebab names, then I definitely agree we should match WIT on casing and just avoid letting people use two case-insensitively equal names in a given context.

**Reply by lukewagner** [reply]
At the moment, namespaces are [`label`s](https://github.com/WebAssembly/component-model/blob/main/design/mvp/Explainer.md#import-and-export-definitions) which are full kebab names.  I didn't put a ton of thought into whether namespaces really needed this; it's just that `label` tends to be our most basic string token used in all the places.

**Comment by calvinrp** [comment]
After a lengthy discussion with @esoterra and @lann, we came to agreement that we need to hash (`LogId`s) and match names in a case sensitive way, but we should enforce uniqueness on both namespaces and package names in a case insensitive way.

Since changing the case of a name affects the generated bindgen code, it can change behavior.

This does open up a possible typo squatting situation if a registry does not correctly enforce case insensitive uniqueness. The federating registry that imports another registry's namespace should check that this uniqueness is enforced.

Registries may choose to adopt a more restrictive policy that only allows lowercase names, for instance.

**Comment by oovm** [comment]
Hi, @calvinrp ,Who does `Registries` refer to in this context?

For example, if I am the owner of the namespace, do I have permission to require additional checks for the style of the uploaded package? The style must be met for successful publishing.

For example, some languages (such as C#) may like upper camel case, and some prefer all lowercase letters.

**Reply by calvinrp** [reply]
@oovm `Registries` refers to any Warg registry instance. During package publish, a registry reject the package for any reason that it chooses. Having pluggable Wasm components that can enforce custom policies is something that we are working towards.

The context of this thread isn't for a custom policy but rather something to enforce in the Wasm component spec and at the registry level. We landed on having package namespaces and names be all lowercase. When you generate bindings, `wit-bindgen` or such, you can convert the identifiers to whatever convention makes sense in the target language.

Related:
https://github.com/WebAssembly/component-model/pull/338
PR merged https://github.com/bytecodealliance/registry/pull/265
